knimin/lib/data_access.py

```diff
@@ -2443,35 +2449,28 @@ def _study_is_unique(self, qiita_study_id=None, title=None, skip_id=None):
         ValueError
             If neither value is given, or either value already exists
         """
-        tags = ['Qiita study ID', 'Title']
-        cols = [tag.lower().replace(' ', '_') for tag in tags]
+        cols = ['qiita_study_id', 'title']
         vals = [vars()[col] for col in cols]
+        tags = [x.capitalize().replace('_', ' ').replace(' id', ' ID')
+                for x in cols]
         if all(val is None for val in vals):
             raise ValueError('Either %s or %s must be specified.'
                              % (tags[0], tags[1]))
-        dups = [0 for tag in tags]
-        for i in range(len(vals)):
-            if tags[i]:
-                sql = """SELECT study_id
-                         FROM pm.study
-                         WHERE {} = %s
-                         AND study_id IS DISTINCT FROM %s""".format(cols[i])
-                sql_args = [vals[i], skip_id]
-                res = self._con.execute_fetchone(sql, sql_args)
-                if res:
-                    dups[i] = res[0]
-        if dups[0] and len(set(dups)) <= 1:
-            raise ValueError('%s %s and %s %s conflict with exisiting study '
-                             '%s.' % (tags[0], repr(vals[0]), tags[1],
-                                      repr(vals[1]), dups[0]))
-        else:
+        with TRN:
+            sql = """SELECT study_id
+                     FROM pm.study
+                     WHERE {} = %s
+                     AND study_id IS DISTINCT FROM %s"""
             errs = []
-            for i in range(len(tags)):
-                if dups[i]:
-                    errs.append('%s %s conflicts with exisiting study %s.'
-                                % (tags[i], repr(vals[i]), dups[i]))
+            err_msg = "%s %s conflicts with exisiting study %s."
+            for col, val, tag in zip(cols, vals, tags):
+                if val:
+                    TRN.add(sql.format(col), [val, skip_id])
+                    res = TRN.execute_fetchflatten()
+                    if res:
+                        errs.append(err_msg % (tag, repr(val), res[0]))
```

Comment:
That's because `repr` will add quotes to a string but not to a number. When applying to ID and title, the outcome will read like: `Warning: Sample ID 123 'my favorite sample' is not valid.`